### PR TITLE
Advanced Search with complex queries 

### PR DESCRIPTION
**Pull request by x5engine**
Fix $or issue and implement the new complex search
https://github.com/blockstack/radiks/issues/67

### CODE REVIEW

File: src/controllers/ModelsController.ts [hunk 1]
```diff
@@ -62,6 +62,25 @@ const makeModelsController = (
     });
   });
 
+  ModelsController.getAsync('/advancedSearch', async (req, res) => {
+    //A custom route for complex search since we can't use queryToMongo
+    const query = req.body;
+    const options = { limit: req.query.limit, offset: req.query.offset };
+    //we keep the same code here but use the radiksCollection
```
Comment (friedger) [inline src/controllers/ModelsController.ts]: Same code to where?

File: src/controllers/ModelsController.ts [hunk 2]
```diff
@@ -62,6 +62,25 @@ const makeModelsController = (
     });
   });
 
+  ModelsController.getAsync('/advancedSearch', async (req, res) => {
+    //A custom route for complex search since we can't use queryToMongo
```
Comment (friedger) [inline src/controllers/ModelsController.ts]: better:    ` //A custom route for complex search that can't be processed by queryToMongo`

File: src/controllers/ModelsController.ts [hunk 3]
```diff
@@ -62,6 +62,25 @@ const makeModelsController = (
...
+    const query = req.body;
+    const options = { limit: req.query.limit, offset: req.query.offset };
+    //we keep the same code here but use the radiksCollection
+    const cursor = radiksCollection.find(query, options);
+    const results = await cursor.toArray();
+    const total = await cursor.count();
+
+    const fullUrl = `${req.protocol}://${req.get('host')}${req.originalUrl}`;
+    const pageLinks = mongo.links(fullUrl.split('?')[0], total);
```
Comment (friedger) [inline src/controllers/ModelsController.ts]: Where is mongo.links defined?

File: src/controllers/ModelsController.ts [hunk 4]
```diff
@@ -62,6 +62,25 @@ const makeModelsController = (
...
+
+    res.json({
+      ...pageLinks,
```
Comment (friedger) [inline src/controllers/ModelsController.ts]: How can pageLinks be used by the client?

File: src/controllers/ModelsController.ts (hunk 3 above)
Comment (x5engine) [reply src/controllers/ModelsController.ts]: Good catch!

I'll add that

File: src/controllers/ModelsController.ts (hunk 4 above)
Comment (x5engine) [reply src/controllers/ModelsController.ts]: We have to make pagination manually then

Comment (friedger) [reply src/controllers/ModelsController.ts]: Some return value to make that easy should be good.

**Comment by x5engine** [comment]
other changes and test with a demo to use the backend feature for this advanced search

https://github.com/x5engine/radiks/pull/2

**Comment by friedger** [comment]
This PR is our solution to https://github.com/blockstack/bounties/issues/3. It consists of this PR and https://github.com/blockstack/radiks/pull/70

It introduces a third parameter to `Model.fetchList`. This parameter expects a JSON object with a `criteria` field describing a database query as defined for mongodb, e.g. for $or see [here](https://docs.mongodb.com/manual/reference/operator/query/or/#op._S_or)

Example:
```
TestModel.fetchList(
    { limit: 1 },
    {},
    { criteria: { $or: [{ name: 'tester' }, {description: {$exists:true}}] } }
  );
```
If the third parameter is present, the `/query` endpoint is used instead of the `/find` endpoint of the radiks server. The `query` endpoint of the `radiks` server accepts a query string like `/find` to define the `limit` parameter. The database query is `POST`ed and send to the mongo database. The endpoint returns the found models in the same format as `/find` and is handled in the same way on the client side.

This solution does not support paging as the current solution does not support paging on the client.

Note, the query object might need to be checked on the server for malicious behavior before sending to the mongo database.

For the pull requests to `radiks` and `radiks-server`, tests have been added. There are is no staging environment available, so no integration tests were added to test the client/server interaction.

**Comment by friedger** [comment]
Work should be continued on #36 
This PR should be closed.

**Pull request closed without merging**